File: 1625-lexicographically-smallest-string-after-applying-operations/1625-lexicographically-smallest-string-after-applying-operations.java
class Solution {
    String ans;
    HashSet<String> set = new HashSet<>();
    int a, b;
    private void dfs(String str){
        if(set.contains(str)) return;
        if(ans.compareTo(str) > 0) ans = str;
        set.add(str);
        dfs(add(str));
        dfs(rotate(str));
    }
    private String add(String str){
        StringBuilder s = new StringBuilder();
        for(int i = 0; i < str.length(); i++){
            if(i % 2 == 0) s.append(str.charAt(i));
            else s.append((str.charAt(i) - '0' + a) %10);
        }
        return s.toString();
    }
    private String rotate(String s){
        int n = s.length();
        return s.substring(n-b, n) + s.substring(0, n-b);
    }
    public String findLexSmallestString(String s, int a, int b) {
        ans = s;
        this.a = a;
        this.b = b;
        dfs(s);
        return ans;
    }
}

File: 1625-lexicographically-smallest-string-after-applying-operations/1625-lexicographically-smallest-string-after-applying-operations.cpp
class Solution {
    unordered_set<string> visited;
    string ans;
    int a, b;
    int n;
    
    bool isVisited(string s) {
        return visited.find(s) != visited.end();
    }
    
    void visit(string s) {
        ans = min(ans, s);
        visited.insert(s);
    }
    
    string rotate(string s, int x) {
        reverse(s.begin(), s.end());
        reverse(s.begin(), s.begin()+x);
        reverse(s.begin()+x, s.end());
        return s;
    }
    
    string add(string s, int x) {
        for (int i=1; i<n; i += 2) {
            char &c = s[i];
            c = '0' + (c-'0'+x)%10;
        }
        return s;
    }
    

public:
    
    void dfs(string s) {
        if (isVisited(s)) {
            return ;
        }
        visit(s);
        dfs(rotate(s, b));
        dfs(add(s, a));
    }
    
    string findLexSmallestString(string s, int a, int b) {
        ans = s;
        this->a = a;
        this->b = b;
        n = s.size();
        dfs(s);
        return ans;
    }
};